src/expressions/method-call-expr.md

```diff
@@ -15,22 +15,168 @@ let log_pi = pi.unwrap_or(1.0).log(2.72);
+
 The following procedure is used:
 
-The first step is to build a list of candidate receiver types.
-Obtain these by repeatedly [dereferencing][dereference] the receiver expression's type, adding each type encountered to the list, then finally attempting an [unsized coercion] at the end, and adding the result type if that is successful.
-Then, for each candidate `T`, add `&T` and `&mut T` to the list immediately after `T`.
+## Determining candidate types
+
+First, a list of "candidate types" is assembled.
+
+These types are found by taking the receiver type and iterating, following either:
+
+* The built-in [dereference]; or
+* `<T as Receiver>::Target`
+
+to the next type. (If a step involved following the `Receiver` target, we also
+note whether it would have been reachable by following `<T as
+Deref::Target>` - this information is used later).
+
+If the final type is an array, an additional candidate type is added
+by [unsized coercion] of that array.
+
+Each step of this chain provides a possible `self` type for methods that
+might be called. The list will be used in two different ways:
+
+* To find types that might have methods. This is used in the
+  "determining candidate methods" step, described below. This considers
+  the full list.
+* To find types to which the receiver can be converted. This is used in the
+  "picking a method from the candidates" step, also described below - in this
+  case, we only consider the types reachable via `Deref` or built-in
+  dereferencing.
+
+There is a built-in implementation of `Receiver` for all `T: Deref`, so
+most of the time, every step can be reached through either mechanism.
+Sometimes, more types can be reached via the `Receiver` chain, and so
+more types will be considered for the former usage than the latter usage.
+
+For instance, if the receiver has type `Box<[i32;2]>`, then the candidate types
+will be `Box<[i32;2]>`,`[i32; 2]` (by dereferencing), and `[i32]` (by unsized
+coercion).
+
+If `SmartPtr<T>: Receiver<Target=T>`, and the receiver type is `&SmartPtr<Foo>`,
+then the candidate types would be `&SmartPtr<Foo>`, `SmartPtr<Foo>` and `Foo`.
+
+## Determining candidate methods
+
+This list of candidate types is then converted to a list of candidate methods.
+For each step, the candidate type is used to determine what searches to perform:
+
+* For a struct, enum, foreign type, or various simpler types (listed below)
+  there is a search for inherent impl candidates for the type.
+* For a type param, there's a search for inherent candidates on the param.
+* For a trait object, there is first a search for inherent candidates for
+  the trait (for example in `impl Trait` blocks), then inherent impl
+  candidates for the trait object itself (for example found in `impl dyn Trait`
+  blocks).
+
+After these occur, there's a further search for extension candidates from
+all traits in scope (irrespective of whether the trait is remotely relevant
+to the `self` type - that's considered later).
```

Comment:
Probably this parenthetical could use some wording iteration.  What I would expect this to mean is "irrespective of whether the trait impl method being considered for selection as a method candidate has a `self` type that's relevant to the receiver type", but not to mean that we're considering methods from traits that aren't relevant to the receiver type due to, e.g., not having any impls for types in the list of candidate types.  Is that right?

Reply:
At this point in the search, we consider _literally any trait_. Limiting to applicable self types comes later in the process.

Reply:
Wild.  You're saying this all gets sorted out during the picking process then.  I suppose that makes sense given that autoref is being handle there also.  OK.